File: Algorithms/p312_Burst_Balloons_1.py
class Solution(object):
    def maxCoins(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        nums_len = len(nums)
        nums.insert(0,1)
        nums.append(1)
        coin = [[0]*(nums_len+2) for i in range(0,nums_len+2)]
        for l in range(0,nums_len):
            for i in range(1,nums_len-l+1):
                j = i+l
                for k in range(i,j+1):
                    coin[i][j] = max(coin[i][j], nums[i-1]*nums[k]*nums[j+1]+coin[i][k-1]+coin[k+1][j])
        return coin[1][nums_len]

File: Algorithms/p312_Burst_Balloons_1.cpp
class Solution {
public:
    int maxCoins(vector<int>& nums) {
        int len = nums.size();
        nums.insert(nums.begin(),1);
        nums.push_back(1);
        vector<vector<int>> coin(len+2,vector<int>(len+2,0));
        for(int l=0;l<len;++l){
            for(int i=1,j=i+l;j<=len;++i,++j){
                int coin_max = 0;
                for(int k=i;k<=j;++k){
                    coin_max = max(coin_max, nums[i-1]*nums[k]*nums[j+1]+coin[i][k-1]+coin[k+1][j]);
                }
                coin[i][j] = coin_max;
            }
        }
        return coin[1][len];
    }
};
